packages/client/src/runtime/DataLoader.ts

```diff
@@ -65,6 +68,7 @@ export class DataLoader<T = unknown> {
             batch[0].reject(e)
           })
       } else {
+        batch.sort((a, b) => this.options.batchOrder(a.request, b.request))
```

Comment:
Is this a symptom of something we need to fix that lives more deeply, or is it the only way to fix it? I am curious to know why/how we could not preserve the order.

Reply:
So, what happens is:
- we call `prisma.$queryRaw` and await it. This triggers `PrismaPromise` and it passes through all the middlewares and gets to a query extension.
- extension wraps it into `$transaction` and adds 2 extra queries, 1 before and 1 after the original. 
- `$transaction` triggers all 3 `PrismaPromises`. Middle promise (original one) is already in progress and does not need pass through middleware. 2 new requests that extension added still do. As a result, middle promise reaches data loader before either of the outer ones.
-  Due to `lock` promise we won't flush the batch until all Promises have reached data loader (this part works correctly), but `DataLoader` still sends them to the engine in the order they come in, which means, middle promise is sent first. 

I think sorting is simplest solution for this case - we guaranteed to have all necessary promises by the time we flush the batch, we know what order they should be. Even simpler solution would've been re-implementing batching to bypass `DataLoader`, but that would require much more time.

Reply:
I think this is a more general problem and is not limited to extensions. This simple code

```ts
prisma.$use(async (params, next) => {
  await new Promise((r) => setTimeout(r, Math.random() * 1000))
  return next(params)
})

await prisma.$transaction([
  prisma.$queryRawUnsafe('SELECT 1'),
  prisma.$queryRawUnsafe('SELECT 2'),
  prisma.$queryRawUnsafe('SELECT 3'),
])
```

Reply:
Another way to fix it would be to use the locks to unlock in the right order, but let's rather remove all of this asap.